src/LibraryManager.Vsix/Shared/LibraryCommandService.cs

```diff
@@ -0,0 +1,352 @@
+                {
+                    Dependencies dependencies = Dependencies.FromConfigFile(configFilePath);
+                    Manifest manifest = await Manifest.FromFileAsync(configFilePath, dependencies, cancellationToken).ConfigureAwait(false);
+                    manifests.Add(configFilePath, manifest);
+                }
+            }
+            catch (Exception ex)
+            {
+                Logger.LogEvent(LibraryManager.Resources.Text.Restore_OperationFailed + Environment.NewLine + ex.Message, LogLevel.Operation);
+            }
+
+            return manifests;
+        }
+
+        private async Task CleanLibrariesAsync(ProjectItem configProjectItem, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            Logger.LogEventsHeader(OperationType.Clean, string.Empty);
+
+            try
+            {
+                Stopwatch sw = new Stopwatch();
+                string configFileName = configProjectItem.FileNames[1];
+                var dependencies = Dependencies.FromConfigFile(configFileName);
+                Manifest manifest = await Manifest.FromFileAsync(configFileName, dependencies, CancellationToken.None).ConfigureAwait(false);
+                IHostInteraction hostInteraction = dependencies.GetHostInteractions();
+
+                sw.Start();
+                IEnumerable<ILibraryOperationResult> results = await manifest.CleanAsync(async (filesPaths) => await hostInteraction.DeleteFilesAsync(filesPaths, cancellationToken), cancellationToken);
+                sw.Stop();
+
+                Logger.LogEventsSummary(results, OperationType.Clean, sw.Elapsed);
+            }
+            catch (OperationCanceledException)
+            {
+                Logger.LogEvent(LibraryManager.Resources.Text.Clean_OperationCancelled, LogLevel.Task);
+            }
+        }
+
+        private async Task RestoreInternalAsync(IDictionary<string, Manifest> manifests, CancellationToken cancellationToken)
+        {
+            Logger.LogEventsHeader(OperationType.Restore, string.Empty);
+
+            Stopwatch sw = new Stopwatch();
+            List<ILibraryOperationResult> totalResults = new List<ILibraryOperationResult>();
+
+            try
+            {
+                sw.Start();
+
+                foreach (KeyValuePair<string, Manifest> manifest in manifests)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    Project project = VsHelpers.GetDTEProjectFromConfig(manifest.Key);
+
+                    Logger.LogEvent(string.Format(LibraryManager.Resources.Text.Restore_LibrariesForProject, project?.Name), LogLevel.Operation);
+
+                    IEnumerable<ILibraryOperationResult> results = await RestoreLibrariesAsync(manifest.Value, cancellationToken).ConfigureAwait(false);
+
+                    await AddFilesToProjectAsync(manifest.Key, project, results, cancellationToken);
```

Comment:
No ConfigureAwait here? Also, assuming the RestoreLibrariesAsync calls are slow, could you do them in parallel?

Reply:
We had this discussion before... RestoreLibrariesAsync "might" do bunch of HttpClient requests if files are not cached. I had to increase the number of "MaxConnectionsPerServer" which improved the performance but there were concerns on the load we generate on the providers. We are still sending requests in parallel for each individual library but sequentially restoring each library and manifests. Thoughts?